src/tools/debugger/mhpdbgc.c

```diff
@@ -556,32 +557,31 @@ static void mhp_stop(int argc, char * argv[])
    }
 }
 
-#define TRACE_STACK_LEN 32 // 16 interrupts deep
-static uint16_t trace_stack[TRACE_STACK_LEN];
-static int trace_stack_index;
+static struct rng_s trace_ringbuf;
 
 static void trace_stack_push(uint16_t seg, uint16_t ofs)
 {
-  assert((trace_stack_index + 2) <= TRACE_STACK_LEN);
+  far_t csip = MK_FARt(seg, ofs);
 
-  trace_stack[trace_stack_index++] = seg;
-  trace_stack[trace_stack_index++] = ofs;
+  rng_put(&trace_ringbuf, &csip);
 }
 
 static void trace_stack_pop(uint16_t *seg, uint16_t *ofs)
 {
-  assert((trace_stack_index - 2) >= 0);
+  far_t csip;
 
-  *ofs = trace_stack[--trace_stack_index];
-  *seg = trace_stack[--trace_stack_index];
+  assert(rng_get(&trace_ringbuf, &csip));
```

Comment:
No-no, this doesn't work!
In non-debug builds (with NDEBUG defined),
the whole `rng_get()` will disappear.
Never ever put any useful code into assert.

Reply:
Wow, I didn't know that! I've never been a fan of `assert()`, I don't like programs to stop as I'd rather they handle the condition. All those malloc() / null pointer check conditions sprinkled about will stop too. I don't expect you'll see me assert again!

Reply:
Asserts are not to handle some condition.
They are to detect an inconsistency in the
program itself, and not let the problem to
propagate to some memory corruption or
something else difficult to debug. It is better
to crash program sooner than later, because
then you have a chance to track the crash
back to something reasonable.
In our case if we returned from an int for
which there is no more a return address,
obviously you don't handle that condition
in any sensible way.
Arguably, as this condition can be provoked
by the guest code, it should be converted
to the clean shutdown.

Reply:
`error()` and `leavedos()` okay?

Reply:
Yes. :)